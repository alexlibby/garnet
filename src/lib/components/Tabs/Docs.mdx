import { Canvas, Meta } from "@storybook/blocks";
import { ArgTypes } from "@storybook/blocks";
import * as TabsStories from "./Tabs.stories";

<Meta of={TabsStories} />

# Tabs

Tabs is the primary component. It has two possible states.

- [Default](#default)
- [Vertical](#vertical)

<a id="default" />

## Default

This is the default version of the Tabs component.

<Canvas of={TabsStories.Default} />

## Vertical

This is the vertical version of the Tabs component.

<Canvas of={TabsStories.Vertical} />

## Properties of component

Below is a list of arguments available for this component:

<ArgTypes of={TabsStories} />
