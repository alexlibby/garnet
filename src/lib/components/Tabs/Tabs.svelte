<svelte:options customElement = "garnet-tabs" />

<script>
  export let activeTabValue = "0";
  export let items = [];
  export let vertical = false;

  if (items == null) {
    items = tabItems;
  }

  const handleClick = tabValue => () => (activeTabValue = tabValue);

  $: vertical = vertical
</script>

<div class="garnet-tabs" class:vertical>
    <ul class:vertical>
      {#each Object.entries(items) as [id]}
        <li class={activeTabValue === id ? 'active' : ''} data-testid="tabHeader">
          <!-- svelte-ignore a11y-click-events-have-key-events -->
          <!-- svelte-ignore a11y-no-static-element-interactions -->
          <span class:vertical on:click={handleClick(id)}>           
            {JSON.stringify(items[id].name).replace(/['"]+/g, "")}
          </span>
        </li>
      {/each}
    </ul>
  
    <div class="content" class:vertical data-testid="tabContent">
      {#each Object.entries(items) as [id]}
        {#if activeTabValue === id}
          {JSON.stringify(items[id].text).replace(/['"]+/g, "")}
        {/if}
      {/each}
    </div>
</div>

<style>
  .garnet-tabs { 
    display: flex; 
    flex-direction: column;
		font-family: Arial, Helvetica, sans-serif;    
  }

  ul {
    display: flex;
    flex-wrap: wrap;
    padding-left: 0;
    margin-bottom: 0;
    list-style: none;
    border-bottom: 1px solid #dee2e6;
  }
	li {
		margin-bottom: -1px;
	}

  span {
    border: 1px solid transparent;
    border-top-left-radius: 4px;
    border-top-right-radius: 4px;
    display: block;
    padding: 8px 16px;
    cursor: pointer;
  }

  span:hover {
    border-color: #dee2e6;
    border-bottom-color: #ffffff;
  }

  li.active > span {
    color: #ffffff;
    background-color: #733635;
    border-color: #dee2e6 #dee2e6 #fff;
  }

  .content {
    padding: 10px;
    min-height: 300px;
  }

  /* variant */
  .garnet-tabs.vertical {
    display: flex;
    flex-direction: row;
  }

  ul.vertical{
    display: flex;
    flex-direction: column;
    border-right: 1px solid #dee2e6;
    border-bottom: none;
    margin-top: 0;
  }

  ul.vertical span {
    border-radius: 0;
    border-right: 0;
    border-color: #dee2e6;
  }
  </style>