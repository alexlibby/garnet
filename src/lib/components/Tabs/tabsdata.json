[
  { "id": 1, "name": "Tab 1", "text": "Text 1" },
  { "id": 2, "name": "Tab 2", "text": "Text 2" },
  { "id": 3, "name": "Tab 3", "text": "Text 3" }
]
