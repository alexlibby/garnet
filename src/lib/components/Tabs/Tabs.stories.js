import Tabs from "./Tabs.svelte";
import tabItems from "./tabsdata.json";
import { BADGE } from "storybook-addon-badges";

const BADGES = {
  ...BADGE,
  ALPHA: "Alpha",
};

export default {
  title: "Garnet UI Library/Navigation Components/Tabs",
  component: Tabs,
  argTypes: {
    vertical: false,
    items: tabItems,
  },
};

export const Default = {
  args: {
    vertical: false,
    items: tabItems,
  },
};

Default.parameters = {
  badges: [BADGES.ALPHA],
  docs: { source: { type: "code" } },
};

export const Vertical = {
  args: {
    vertical: true,
    items: tabItems,
  },
};

Vertical.parameters = {
  badges: [BADGES.ALPHA],
  docs: { source: { type: "code" } },
};
