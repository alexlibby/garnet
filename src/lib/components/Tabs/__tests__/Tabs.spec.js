import { describe, test, expect } from "vitest";
import { axe } from "vitest-axe";
import * as AxeMatchers from "vitest-axe/matchers";
import { render, screen, fireEvent } from "@testing-library/svelte";
import { userEvent } from "@vitest/browser/context";
import Tabs from "../../Tabs/Tabs.svelte";

const TabsHTML = `<div class="garnet-tabs s-NPWaW4eOvB4v"><div class="container s-NPWaW4eOvB4v"><ul class="s-NPWaW4eOvB4v"><li class="active s-NPWaW4eOvB4v" data-testid="tabHeader"><span class="s-NPWaW4eOvB4v">Tab 1</span> </li><li class=" s-NPWaW4eOvB4v" data-testid="tabHeader"><span class="s-NPWaW4eOvB4v">Tab 2</span> </li><li class=" s-NPWaW4eOvB4v" data-testid="tabHeader"><span class="s-NPWaW4eOvB4v">Tab 3</span> </li></ul> <div class="content s-NPWaW4eOvB4v" data-testid="tabContent">Text 1</div></div></div>`;
expect.extend(AxeMatchers);

describe("Tests for Tabs", () => {
  const props = {
    vertical: false,
    items: [
      { id: 1, name: "Tab 1", text: "This is a test" },
      { id: 2, name: "Tab 2", text: "Here is tab 2" },
      { id: 3, name: "Tab 3", text: "And this is tab 3" },
    ],
  };

  test("should render properly", () => {
    const result = render(Tabs, props);
    expect(() => result).not.toThrow();
  });

  test("should change text if clicking on tab 2", async () => {
    render(Tabs, props);
    const secondTabElement = screen.queryByText("Tab 2");
    await fireEvent.click(secondTabElement);

    const secondTabElementText = await screen.findByText(/Here is tab 2/i);
    expect(secondTabElementText).toBeTruthy();
  });

  test("get a snapshot of component", () => {
    // Create a new container for the test and mount component
    const host = document.createElement("div");
    document.body.appendChild(host);
    const instance = new Tabs({ target: host, props: props });

    // check that all rendered OK
    expect(instance).toBeTruthy();

    // Take snapshot
    expect(host.innerHTML).toMatchSnapshot();
  });

  test("should demonstrate no issues with accessibility", async () => {
    const render = () => TabsHTML;
    const result = render();

    const results = await axe(result, {
      rules: {
        region: { enabled: false },
      },
    });

    expect(results).toHaveNoViolations();
  });
});
