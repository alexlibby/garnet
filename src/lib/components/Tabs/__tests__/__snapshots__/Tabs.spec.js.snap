// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Tests for Tabs > get a snapshot of component 1`] = `"<div class="garnet-tabs svelte-1dkyasc"><ul class="svelte-1dkyasc"><li class="active svelte-1dkyasc" data-testid="tabHeader"><span class="svelte-1dkyasc">Tab 1</span> </li><li class=" svelte-1dkyasc" data-testid="tabHeader"><span class="svelte-1dkyasc">Tab 2</span> </li><li class=" svelte-1dkyasc" data-testid="tabHeader"><span class="svelte-1dkyasc">Tab 3</span> </li></ul> <div class="content svelte-1dkyasc" data-testid="tabContent">This is a test</div></div>"`;
